apis/placement/v1alpha1/stagerollout_types.go

```diff
@@ -0,0 +1,299 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope="Namespaced",categories={fleet,fleet-placement}
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// StageRollout defines a stage by stage rollout policy that is applied to the ClusterResourcePlacement it tracks.
+type StageRollout struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// The desired state of StageRollout.
+	// +required
+	Spec StageRolloutSpec `json:"spec"`
+
+	// The observed status of StageRollout.
+	// +optional
+	Status StageRolloutStatus `json:"status,omitempty"`
+}
+
+// StageRolloutSpec defines the desired the rollout sequence stage by stage.
+type StageRolloutSpec struct {
+	// The name of the CRP that this rollout is applied to.
+	// There can only be one active rollout for each ClusterResourcePlacement,
+	// we will fail the rollout if there is already an active rollout.
+	// +required
+	ParentCRPName string `json:"parentCRPName"`
+
+	// The resource snapshot index that this rollout is based on.
+	// The rollout will use the latest resource snapshot if this is empty.
+	// +optional
+	ResourceSnapShotIndex string `json:"resourceSnapShotIndex,omitempty"`
+
+	// Stage rollout configurations for each rollout stage.
+	// +required
+	Stages []StageRolloutConfig `json:"stages"`
+
+	// The common post rollout tasks that ALL stages needs to be completed before moving to the next stage.
+	// +optional
+	PostRolloutTasks PostRolloutTask `json:"postRolloutTasks,omitempty"`
+}
+
+// StageRolloutConfig describes a single rollout stage group configuration.
+type StageRolloutConfig struct {
```

Comment:
how do we rollout the selected clusters within a stage?

sequentially or concurrently?